### PR TITLE
drivers/dma: support dma driver model

### PR DESCRIPTION
**Pull request by Donny9**
## Summary
drivers/dma: support dma driver model
1. support dma controller interface.
2. support dma link.
## Impact
support common dma ip
## Testing
local test

### CODE REVIEW

File: include/nuttx/dma/dma.h [hunk 1]
```diff
@@ -0,0 +1,313 @@
...
+    unsigned int link_size;
+  };
+};
+
+struct dma_link_config_s
+{
+  unsigned int dst_link_num;
+  unsigned int src_link_num;
+  struct dma_link_s *dst_link;
+  struct dma_link_s *src_link;
```
Comment (pkarashchenko) [inline include/nuttx/dma/dma.h:267]:
```suggestion
  FAR struct dma_link_s *dst_link;
  FAR struct dma_link_s *src_link;
```

Comment (Donny9) [reply include/nuttx/dma/dma.h]: Done, Tnks~

File: include/nuttx/dma/dma.h [hunk 2]
```diff
@@ -0,0 +1,313 @@
...
+#include <stdint.h>
+#include <sys/types.h>
+
+/****************************************************************************
+ * Pre-processor Definitions
+ ****************************************************************************/
+
+/* DMA transfer direction indicator */
+
+#define DMA_MEM_TO_MEM          1
```
Comment (pkarashchenko) [inline include/nuttx/dma/dma.h:38]: Just a question: If SoC support specific DMA that can transfer only SRAM to PSRAM copy and vice versa, so is usage of that DMA covered by `DMA_MEM_TO_MEM`?

**Review by pkarashchenko — APPROVED** [review]

File: include/nuttx/dma/dma.h [hunk 3]
```diff
@@ -0,0 +1,313 @@
...
+#define DMA_MEM_TO_DEV          2
+#define DMA_DEV_TO_MEM          3
+#define DMA_DEV_TO_DEV          4
+
+#ifdef CONFIG_DMA_LINK
+# define DMA_BLOCK_MODE         0
+# define DMA_SRC_LINK_MODE      1
+# define DMA_DST_LINK_MODE      2
+# define DMA_DUAL_LINK_MODE     3
```
Comment (pkarashchenko) [inline include/nuttx/dma/dma.h:47]:
```suggestion
#  define DMA_BLOCK_MODE        0
#  define DMA_SRC_LINK_MODE     1
#  define DMA_DST_LINK_MODE     2
#  define DMA_DUAL_LINK_MODE    3
```

File: include/nuttx/dma/dma.h (hunk 2 above)
Comment (xiaoxiang781216) [reply include/nuttx/dma/dma.h:38]: Yes

**Review by pkarashchenko — APPROVED** [review]

**Comment by Donny9** [comment]
@pkarashchenko please merge this driver model, thank you !

File: include/nuttx/dma/dma.h [hunk 4]
```diff
@@ -0,0 +1,313 @@
...
+struct dma_ops_s
+{
+  int (*config)(FAR struct dma_chan_s *chan,
+                FAR const struct dma_config_s *cfg);
+  int (*start)(FAR struct dma_chan_s *chan,
+               dma_callback_t callback, FAR void *arg,
+               uintptr_t dst, uintptr_t src, size_t len);
+  int (*start_cyclic)(FAR struct dma_chan_s *chan,
+                      dma_callback_t callback, FAR void *arg,
+                      uintptr_t dst, uintptr_t src,
```
Comment (pkarashchenko) [inline include/nuttx/dma/dma.h:282]: Please add `CODE` for function pointers

Comment (Donny9) [reply include/nuttx/dma/dma.h]: Done, thank you.

File: include/nuttx/dma/dma.h [hunk 5]
```diff
@@ -0,0 +1,318 @@
...
+ *   len  - Positive value represent the transfer length, negative is error
+ *          code.
+ *
+ * Returned Value:
+ *   None
+ *
+ ****************************************************************************/
+
+struct dma_chan_s;
+typedef void (*dma_callback_t)(FAR struct dma_chan_s *chan, FAR void *arg,
```
Comment (pkarashchenko) [inline include/nuttx/dma/dma.h:217]:
```suggestion
typedef CODE void (*dma_callback_t)(FAR struct dma_chan_s *chan, FAR void *arg,
```

Comment (Donny9) [reply include/nuttx/dma/dma.h]: Done!

**Review by pkarashchenko — APPROVED** [review]

**Comment by linguini1** [comment]
Hi, I'm interested in using this as part of my port for the Pi4B, but it looks like the documentation for it is blank: https://nuttx.apache.org/docs/latest/components/drivers/special/dma.html

Any chance you'd be willing to fill out that documentation page?

**Pull request merged**